File: src/convention/path.ts
import type { ServiceIR, OperationDecl, ConventionsDecl, Expr } from "#ir/types.js";
import type { OperationClassification, EndpointSpec, ParamSpec, HttpMethod } from "#convention/types.js";
import { toPathSegment, toKebabCase } from "#convention/naming.js";

export function deriveEndpoints(
  classifications: readonly OperationClassification[],
  ir: ServiceIR,
): readonly EndpointSpec[] {
  return classifications.map((c) => {
    const op = ir.operations.find((o) => o.name === c.operationName)!;
    return deriveEndpoint(c, op, ir);
  });
}

function deriveEndpoint(
  classification: OperationClassification,
  op: OperationDecl,
  ir: ServiceIR,
): EndpointSpec {
  const method = resolveMethod(classification, ir.conventions);
  const path = resolvePath(classification, op, ir);
  const successStatus = resolveStatus(classification, ir.conventions, method);

  const pathParamNames = extractPathParamNames(path);
  const pathParams: ParamSpec[] = [];
  const otherParams: ParamSpec[] = [];

  for (const input of op.inputs) {
    if (pathParamNames.has(input.name)) {
      pathParams.push({
        name: input.name,
        typeExpr: input.typeExpr,
        required: true,
      });
    } else {
      otherParams.push({
        name: input.name,
        typeExpr: input.typeExpr,
        required: input.typeExpr.kind !== "OptionType",
      });
    }
  }

  const isGet = method === "GET";
  return {
    operationName: classification.operationName,
    method,
    path,
    pathParams,
    queryParams: isGet ? otherParams : [],
    bodyParams: isGet ? [] : otherParams,
    successStatus,
  };
}

const VALID_METHODS = new Set<HttpMethod>(["GET", "POST", "PUT", "PATCH", "DELETE"]);

function resolveMethod(
  classification: OperationClassification,
  conventions: ConventionsDecl | null,
): HttpMethod {
  const override = getConvention(conventions, classification.operationName, "http_method");
  if (override && VALID_METHODS.has(override as HttpMethod)) {
    return override as HttpMethod;
  }
  return classification.method;
}

function resolvePath(
  classification: OperationClassification,
  op: OperationDecl,
  ir: ServiceIR,
): string {
  const override = getConvention(ir.conventions, op.name, "http_path");
  if (override) return override;

  return autoDerivePath(classification, op, ir);
}

function autoDerivePath(
  classification: OperationClassification,
  op: OperationDecl,
  ir: ServiceIR,
): string {
  const entity = classification.targetEntity;
  const segment = entity ? toPathSegment(entity) : toKebabCase(op.name);

  switch (classification.kind) {
    case "create":
      return `/${segment}`;

    case "read":
    case "filtered_read": {
      const idParam = findIdParam(op, ir);
      if (idParam) return `/${segment}/{${idParam}}`;
      return `/${segment}`;
    }

    case "replace":
    case "partial_update": {
      const idParam = findIdParam(op, ir);
      if (idParam) return `/${segment}/{${idParam}}`;
      return `/${segment}`;
    }

    case "delete": {
      const idParam = findIdParam(op, ir);
      if (idParam) return `/${segment}/{${idParam}}`;
      return `/${segment}`;
    }

    case "transition": {
      const idParam = findIdParam(op, ir);
      const action = extractActionVerb(op.name, entity);
      if (idParam) return `/${segment}/{${idParam}}/${action}`;
      return `/${segment}/${action}`;
    }

    case "batch_mutation":
      return `/${segment}/batch`;

    case "side_effect":
      return `/${toKebabCase(op.name)}`;

    case "create_child":
      return `/${segment}`;
  }
}

function findIdParam(op: OperationDecl, ir: ServiceIR): string | null {
  if (!ir.state) return null;

  const keyTypeNames = new Set<string>();
  for (const field of ir.state.fields) {
    if (field.typeExpr.kind === "RelationType") {
      const keyName = typeExprName(field.typeExpr.fromType);
      if (keyName) keyTypeNames.add(keyName);
    }
  }

  for (const input of op.inputs) {
    const inputTypeName = typeExprName(input.typeExpr);
    if (inputTypeName && keyTypeNames.has(inputTypeName)) {
      return input.name;
    }
    if (input.typeExpr.kind === "NamedType" && input.typeExpr.name === "Int") {
      if (input.name === "id" || input.name.endsWith("_id")) {
        return input.name;
      }
    }
  }

  return null;
}

function typeExprName(typeExpr: { kind: string; name?: string }): string | null {
  if (typeExpr.kind === "NamedType" && "name" in typeExpr) {
    return typeExpr.name as string;
  }
  return null;
}

function extractActionVerb(opName: string, entityName: string | null): string {
  if (!entityName) return toKebabCase(opName);

  if (opName.endsWith(entityName)) {
    const verb = opName.slice(0, -entityName.length);
    if (verb.length > 0) return toKebabCase(verb);
  }
  if (opName.startsWith(entityName)) {
    const verb = opName.slice(entityName.length);
    if (verb.length > 0) return toKebabCase(verb);
  }

  return toKebabCase(opName);
}

function extractPathParamNames(path: string): Set<string> {
  const result = new Set<string>();
  const regex = /\{(\w+)\}/g;
  let match;
  while ((match = regex.exec(path)) !== null) {
    result.add(match[1]);
  }
  return result;
}

function resolveStatus(
  classification: OperationClassification,
  conventions: ConventionsDecl | null,
  effectiveMethod: HttpMethod,
): number {
  const override = getConvention(conventions, classification.operationName, "http_status_success");
  if (override) return Number(override);

  if (effectiveMethod === "DELETE") return 204;

  switch (classification.kind) {
    case "create":
    case "create_child":
      return 201;
    case "delete":
      return 204;
    default:
      return 200;
  }
}

export function getConvention(
  conventions: ConventionsDecl | null,
  target: string,
  property: string,
): string | null {
  if (!conventions) return null;

  for (const rule of conventions.rules) {
    if (rule.target === target && rule.property === property) {
      return exprToString(rule.value);
    }
  }
  return null;
}

function exprToString(expr: Expr): string | null {
  switch (expr.kind) {
    case "StringLit":
      return expr.value;
    case "IntLit":
      return String(expr.value);
    case "FloatLit":
      return String(expr.value);
    case "BoolLit":
      return String(expr.value);
    default:
      return null;
  }
}

File: modules/convention/src/main/scala/specrest/convention/Path.scala
package specrest.convention

import specrest.ir.*
import specrest.ir.generated.SpecRestGenerated
import specrest.ir.generated.SpecRestGenerated.*

object Path:

  private given CanEqual[http_method, http_method] = CanEqual.derived

  def deriveEndpoints(
      classifications: List[operation_classification],
      ir: ServiceIRFull
  ): List[EndpointSpec] =
    classifications.map: c =>
      val opName = classificationOperationName(c)
      val op = svcOperations(ir)
        .find(o => operName(o) == opName)
        .getOrElse(
          throw new RuntimeException(s"operation not found: $opName")
        )
      deriveEndpoint(c, op, ir)

  private def deriveEndpoint(
      classification: operation_classification,
      op: operation_decl_full,
      ir: ServiceIRFull
  ): EndpointSpec =
    val method        = resolveMethod(classification, svcConventions(ir))
    val path          = resolvePath(classification, op, ir)
    val successStatus = resolveStatus(classification, svcConventions(ir), method)

    val pathParamNames = extractPathParamNames(path)
    val pathParams     = List.newBuilder[ParamSpec]
    val other          = List.newBuilder[ParamSpec]
    for p <- operInputs(op) do
      val name = prmName(p)
      val ty   = prmType(p)
      if pathParamNames.contains(name) then
        pathParams += ParamSpec(name, ty, required = true)
      else
        val required: Boolean =
          ty match
            case _: OptionTypeF => false
            case _              => true
        other += ParamSpec(name, ty, required)

    val isGet = isGetMethod(method)
    EndpointSpec(
      operationName = classificationOperationName(classification),
      method = method,
      path = path,
      pathParams = pathParams.result(),
      queryParams = if isGet then other.result() else Nil,
      bodyParams = if isGet then Nil else other.result(),
      successStatus = successStatus
    )

  private def resolveMethod(
      c: operation_classification,
      conv: Option[conventions_decl_full]
  ): http_method =
    val override_ = getConvention(conv, classificationOperationName(c), "http_method")
      .flatMap(parseHttpMethod)
    SpecRestGenerated.resolveMethod(override_, classificationMethod(c))

  private def resolvePath(
      c: operation_classification,
      op: operation_decl_full,
      ir: ServiceIRFull
  ): String =
    getConvention(svcConventions(ir), operName(op), "http_path")
      .getOrElse(autoDerivePath(c, op, ir))

  private def autoDerivePath(
      c: operation_classification,
      op: operation_decl_full,
      ir: ServiceIRFull
  ): String =
    val entity  = classificationTargetEntity(c)
    val opKebab = Naming.toKebabCase(operName(op))
    val segment = entity.map(Naming.toPathSegment).getOrElse(opKebab)
    val action  = extractActionVerb(operName(op), entity)
    val idOpt   = findIdParam(op, ir)
    derivePathPattern(classificationKind(c), segment, idOpt, action, opKebab)

  private def findIdParam(op: operation_decl_full, ir: ServiceIRFull): Option[String] =
    SpecRestGenerated.findIdParam(operInputs(op), svcState(ir))

  private def extractActionVerb(opName: String, entityName: Option[String]): String =
    Naming.toKebabCase(extractVerbBeforeKebab(opName, entityName))

  private val PathParamRegex = """\{(\w+)\}""".r

  private def extractPathParamNames(path: String): Set[String] =
    PathParamRegex.findAllMatchIn(path).map(_.group(1)).toSet

  private def resolveStatus(
      c: operation_classification,
      conv: Option[conventions_decl_full],
      effective: http_method
  ): Int =
    val overrideStr = getConvention(conv, classificationOperationName(c), "http_status_success")
    SpecRestGenerated.resolveStatus(overrideStr, effective, classificationKind(c)).toInt

  def getConvention(
      conv: Option[conventions_decl_full],
      target: String,
      property: String
  ): Option[String] =
    conv.flatMap { c =>
      cvdRules(c).collectFirst {
        case ConventionRuleFull(t, p, _, CvOk(pv), _) if t == target && p == property =>
          parsedValueToString(pv)
      }.flatten
    }
